packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java

```diff
@@ -0,0 +1,147 @@
+
+    Assert.assertEquals(1, updatesToDelete.size());
+    Assert.assertTrue(updatesToDelete.contains(updateDefault1));
+    Assert.assertFalse(updatesToDelete.contains(updateDefault2));
+    Assert.assertFalse(updatesToDelete.contains(updateRollout2));
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NormalCase_NewUpdate() {
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateRollout2, updateRollout1, manifestFilters);
+    Assert.assertTrue(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NormalCase_NoUpdate() {
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateRollout1, updateRollout1, manifestFilters);
+    Assert.assertFalse(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NoneMatchingFilters() {
+    // should choose to load an older update if the current update doesn't match the manifest filters
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateRollout1, updateDefault2, manifestFilters);
+    Assert.assertTrue(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NewerExists() {
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateRollout1, updateRollout2, manifestFilters);
+    Assert.assertFalse(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_OlderExistsMatchingFilters() {
+    // should not choose to load a new update that doesn't match the manifest filters
+    // if there is an existing older update that matches the filters
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateDefault2, updateRollout1, manifestFilters);
+    Assert.assertFalse(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NoneExistsMatchingCurrentFilters() {
+    // should choose to load a new update that doesn't match the manifest filters
+    // if no existing updates match the manifest filters
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateDefault2, null, manifestFilters);
+    Assert.assertTrue(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NoneExistsMatchingNewFilters() {
+    // should choose to load a new update that doesn't match the manifest filters
+    // if no existing updates match the manifest filters
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateDefault2, updateDefault1, manifestFilters);
+    Assert.assertTrue(actual);
+  }
+
+  @Test
+  public void testIsUpdateManifestFiltered_Nested() throws JSONException {
+    JSONObject nestedManifestFilters = new JSONObject("{\"nested.object.key\": \"different-value\"}");
+
+    // an update that has the matching key with a different value should be filtered out
+    Assert.assertTrue(selectionPolicy.isUpdateManifestFiltered(updateNested, nestedManifestFilters));
+
+    // an update that doesn't have the key should not be filtered out
+    Assert.assertFalse(selectionPolicy.isUpdateManifestFiltered(updateDefault1, nestedManifestFilters));
+  }
+}
```

Comment:
May also want to add a test that shows that if a manifest doesn't have a field the filter looks for, it's implicitly counted as a match. (That said, do we want that behavior? If a developer says to match a given field and then they send down a manifest that omits the field altogether, I don't think they'd expect it to run.)

Reply:
Agree about adding a test. I'm pretty sure we do want this behavior; otherwise we'd need to start putting a `branchName` (or whatever other filters a user creates) into all embedded manifests in order for them not to be immediately filtered out.

(To be clear, I think we do want to allow users to specify `branchName` and other fields in embedded manifests, but we also don't want to force them to, which is what this "empty matches all" behavior allows.)